includes/validation/class-amp-css-length.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * Class AMP_CSS_Length
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_CSS_Length
+ *
+ * Adapted from the <code>amp.validator.CssLength</code> class found in validator.js of the
+ * <a href="https://github.com/ampproject/amphtml">ampproject/amphtml</a> project.
+ *
+ * @version 1911070201440
+ * @link https://github.com/ampproject/amphtml/blob/1911070201440/validator/engine/validator.js#L3351
+ *
+ * @since 1.5
+ */
+class AMP_CSS_Length {
+
+	/**
+	 * Whether the value or unit is invalid. Note that passing
+	 * an empty value as `$attr_value` is considered valid.
+	 *
+	 * @var bool
+	 */
+	protected $is_valid = false;
+
+	/**
+	 * Whether the attribute value is set.
+	 *
+	 * @var bool
+	 */
+	protected $is_set = false;
+
+	/**
+	 * Whether the attribute value is 'auto'. This is a special value that
+	 * indicates that the value gets derived from the context. In practice
+	 * that's only ever the case for a width.
+	 *
+	 * @var bool
+	 */
+	protected $is_auto = false;
+
+	/**
+	 * Whether the attribute value is 'fluid'.
+	 *
+	 * @var bool
+	 */
+	protected $is_fluid = false;
+
+	/**
+	 * The numeric value.
+	 *
+	 * @var float
+	 */
+	protected $numeral = 0;
+
+	/**
+	 * The unit, 'px' being the default in case it's absent.
+	 *
+	 * @var string
+	 */
+	protected $unit = 'px';
+
+	/**
+	 * AMP_CSS_Length constructor.
+	 *
+	 * @param string $attr_value  Attribute value to be parsed.
+	 * @param bool   $allow_auto  Whether or not to allow the 'auto' value as a value.
+	 * @param bool   $allow_fluid Whether or not to allow the 'fluid' value as a value.
+	 */
+	public function __construct( $attr_value, $allow_auto, $allow_fluid ) {
```

Comment:
I don't think that passing `$allow_auto` and `$allow_fluid` should be done here. For one thing, it makes the instantiations hard to read, for example:

```php
		$input_height = new AMP_CSS_Length(
			$node->getAttribute( 'height' ),
			true,
			true
		);
```

Secondly, it feels like it's somewhat overloading the constructor by doing both capturing the value while also performing validation.

There is validation is being done in the `AMP_Tag_And_Attribute_Sanitizer::has_valid_dimensions()` method, which I think should take more of the responsibility than what is being done in the constructor here.

Reply:
This was a verbatim of [AMP's validator](https://github.com/ampproject/amphtml/blob/1911070201440/validator/engine/validator.js#L3351), but you do make valid points. Will refactor this.